File: excel-report/scripts/create_report.js
#!/usr/bin/env node
/**
 * Bundled Excel report builder for the excel-report skill.
 * Handles logo placement, header layout, and data formatting.
 *
 * Usage:
 *   node create_report.js \
 *     --title "Report Title" \
 *     --summary "Summary sentence" \
 *     --data /tmp/data.json \
 *     --logo resources/company_logo.png \
 *     --output output/report_20250115_120000.xlsx
 */

const ExcelJS = require('exceljs');
const fs = require('fs');
const path = require('path');

const HEADER_BG = '1F4E79';
const HEADER_FG = 'FFFFFFFF';
const ALT_ROW_BG = 'FFD6E4F0';
const SUMMARY_BG = 'FFF2F2F2';
const SUMMARY_FG = 'FF595959';
const TITLE_COLOR = 'FF1F4E79';

const CURRENCY_KEYWORDS = ['salary', 'pay', 'wage', 'compensation', 'income', 'bonus', 'rate'];

function isCurrencyCol(key) {
  return CURRENCY_KEYWORDS.some(kw => key.toLowerCase().includes(kw));
}

function parseArgs(argv) {
  const args = {};
  for (let i = 2; i < argv.length; i += 2) {
    args[argv[i].replace('--', '')] = argv[i + 1];
  }
  return args;
}

async function createReport(title, summary, data, logoPath, outputPath) {
  const workbook = new ExcelJS.Workbook();
  const ws = workbook.addWorksheet('Report');

  // --- Logo + title area (rows 1-2) ---
  ws.getRow(1).height = 45;
  ws.getRow(2).height = 20;
  ws.getColumn('A').width = 4;
  ws.getColumn('B').width = 16;
  ws.getColumn('C').width = 5;

  if (logoPath && fs.existsSync(logoPath)) {
    const ext = path.extname(logoPath).slice(1).toLowerCase();
    const imageId = workbook.addImage({
      filename: logoPath,
      extension: ext === 'jpg' ? 'jpeg' : ext,
    });
    ws.addImage(imageId, {
      tl: { col: 0, row: 0 },
      br: { col: 2, row: 2 },
      editAs: 'oneCell',
    });
  }

  const titleCell = ws.getCell('C1');
  titleCell.value = title;
  titleCell.font = { bold: true, size: 14, color: { argb: TITLE_COLOR } };
  titleCell.alignment = { vertical: 'middle' };

  // --- Summary row (row 3) ---
  ws.getRow(3).height = 18;
  const nCols = data.length > 0 ? Object.keys(data[0]).length : 4;
  const endCol = Math.max(nCols, 4);

  ws.mergeCells(3, 1, 3, endCol);
  const summaryCell = ws.getCell('A3');
  summaryCell.value = summary;
  summaryCell.font = { italic: true, size: 10, color: { argb: SUMMARY_FG } };
  summaryCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: SUMMARY_BG } };
  summaryCell.alignment = { horizontal: 'left', vertical: 'middle', indent: 1 };

  if (data.length === 0) {
    const dir = path.dirname(outputPath);
    if (dir) fs.mkdirSync(dir, { recursive: true });
    await workbook.xlsx.writeFile(outputPath);
    console.log(`Report saved: ${outputPath}`);
    return;
  }

  const headers = Object.keys(data[0]);

  // --- Column headers (row 4) ---
  ws.getRow(4).height = 22;
  headers.forEach((key, i) => {
    const cell = ws.getCell(4, i + 1);
    cell.value = key.replace(/_/g, ' ').replace(/\b\w/g, c => c.toUpperCase());
    cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF' + HEADER_BG } };
    cell.font = { bold: true, color: { argb: HEADER_FG }, size: 11 };
    cell.alignment = { horizontal: 'center', vertical: 'middle' };
  });

  // --- Data rows (row 5+) ---
  data.forEach((record, rowOffset) => {
    const rowNum = rowOffset + 5;
    ws.getRow(rowNum).height = 16;
    headers.forEach((key, colOffset) => {
      const cell = ws.getCell(rowNum, colOffset + 1);
      const raw = record[key];

      if (isCurrencyCol(key)) {
        const num = parseFloat(String(raw).replace(/[$,]/g, ''));
        if (!isNaN(num)) {
          cell.value = num;
          cell.numFmt = '$#,##0';
        } else {
          cell.value = raw;
        }
      } else {
        cell.value = raw;
      }

      if (rowNum % 2 === 0) {
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: ALT_ROW_BG } };
      }
      cell.alignment = { vertical: 'middle' };
    });
  });

  // --- Auto-fit column widths ---
  headers.forEach((key, i) => {
    const col = ws.getColumn(i + 1);
    const headerLabel = key.replace(/_/g, ' ').replace(/\b\w/g, c => c.toUpperCase());
    let maxLen = headerLabel.length;
    data.forEach(record => {
      const val = record[key];
      if (val != null) maxLen = Math.max(maxLen, String(val).length);
    });
    col.width = Math.min(maxLen + 4, 40);
  });

  const dir = path.dirname(outputPath);
  if (dir) fs.mkdirSync(dir, { recursive: true });
  await workbook.xlsx.writeFile(outputPath);
  console.log(`Report saved: ${outputPath}`);
}

(async () => {
  const args = parseArgs(process.argv);
  if (!args.title || !args.summary || !args.data || !args.output) {
    console.error('Usage: node create_report.js --title "..." --summary "..." --data data.json --logo logo.png --output out.xlsx');
    process.exit(1);
  }
  const data = JSON.parse(fs.readFileSync(args.data, 'utf8'));
  await createReport(args.title, args.summary, data, args.logo || 'resources/company_logo.png', args.output);
})();

File: excel-report/scripts/create_report.py
#!/usr/bin/env python3
"""
Bundled Excel report builder for the excel-report skill.
Handles logo placement, header layout, and data formatting.

Usage:
    python create_report.py \
        --title "Report Title" \
        --summary "Summary sentence" \
        --data /tmp/data.json \
        --logo resources/company_logo.png \
        --output output/report_20250115_120000.xlsx
"""

import argparse
import json
import os
from datetime import datetime

from openpyxl import Workbook
from openpyxl.drawing.image import Image
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter

HEADER_BG = "1F4E79"
HEADER_FG = "FFFFFF"
ALT_ROW_BG = "D6E4F0"
SUMMARY_BG = "F2F2F2"
SUMMARY_FG = "595959"

CURRENCY_KEYWORDS = {"salary", "pay", "wage", "compensation", "income", "bonus", "rate"}


def _is_currency_col(key: str) -> bool:
    return any(kw in key.lower() for kw in CURRENCY_KEYWORDS)


def create_report(title: str, summary: str, data: list, logo_path: str, output_path: str):
    wb = Workbook()
    ws = wb.active
    ws.title = "Report"

    # --- Logo + title area (rows 1-2) ---
    ws.row_dimensions[1].height = 35
    ws.row_dimensions[2].height = 20
    ws.column_dimensions["A"].width = 4   # narrow col for logo left margin
    ws.column_dimensions["B"].width = 16
    ws.column_dimensions["C"].width = 5   # spacer between logo and title

    if logo_path and os.path.exists(logo_path):
        img = Image(logo_path)
        img.width = 100
        img.height = 55
        img.anchor = "A1"
        ws.add_image(img)

    title_cell = ws["D1"]
    title_cell.value = title
    title_cell.font = Font(bold=True, size=14, color=HEADER_BG)
    title_cell.alignment = Alignment(vertical="center", wrap_text=False)

    # --- Summary row (row 3) ---
    ws.row_dimensions[3].height = 18
    n_cols = len(data[0]) if data else 4
    end_col = get_column_letter(max(n_cols, 4))
    ws.merge_cells(f"A3:{end_col}3")
    summary_cell = ws["A3"]
    summary_cell.value = summary
    summary_cell.font = Font(italic=True, size=10, color=SUMMARY_FG)
    summary_cell.fill = PatternFill("solid", fgColor=SUMMARY_BG)
    summary_cell.alignment = Alignment(horizontal="left", vertical="center", indent=1)

    if not data:
        os.makedirs(os.path.dirname(output_path) or ".", exist_ok=True)
        wb.save(output_path)
        print(f"Report saved: {output_path}")
        return

    headers = list(data[0].keys())

    # --- Column headers (row 4) ---
    ws.row_dimensions[4].height = 22
    header_fill = PatternFill("solid", fgColor=HEADER_BG)
    header_font = Font(bold=True, color=HEADER_FG, size=11)

    for col_idx, key in enumerate(headers, start=1):
        cell = ws.cell(row=4, column=col_idx)
        cell.value = key.replace("_", " ").title()
        cell.fill = header_fill
        cell.font = header_font
        cell.alignment = Alignment(horizontal="center", vertical="center")

    # --- Data rows (row 5+) ---
    alt_fill = PatternFill("solid", fgColor=ALT_ROW_BG)

    for row_idx, record in enumerate(data, start=5):
        ws.row_dimensions[row_idx].height = 16
        for col_idx, key in enumerate(headers, start=1):
            cell = ws.cell(row=row_idx, column=col_idx)
            value = record.get(key, "")

            if _is_currency_col(key):
                try:
                    cell.value = float(str(value).replace(",", "").replace("$", ""))
                    cell.number_format = '$#,##0'
                except (ValueError, TypeError):
                    cell.value = value
            else:
                cell.value = value

            if row_idx % 2 == 0:
                cell.fill = alt_fill
            cell.alignment = Alignment(vertical="center")

    # --- Auto-fit column widths ---
    for col_idx, key in enumerate(headers, start=1):
        col_letter = get_column_letter(col_idx)
        max_len = len(key.replace("_", " ").title())
        for row_idx in range(5, len(data) + 5):
            val = ws.cell(row=row_idx, column=col_idx).value
            if val is not None:
                max_len = max(max_len, len(str(val)))
        ws.column_dimensions[col_letter].width = min(max_len + 4, 40)

    os.makedirs(os.path.dirname(output_path) or ".", exist_ok=True)
    wb.save(output_path)
    print(f"Report saved: {output_path}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Generate formatted Excel report")
    parser.add_argument("--title", required=True, help="Report title shown beside logo")
    parser.add_argument("--summary", required=True, help="One-sentence summary row")
    parser.add_argument("--data", required=True, help="Path to JSON file (list of dicts)")
    parser.add_argument("--logo", default="resources/company_logo.png", help="Path to logo PNG")
    parser.add_argument("--output", required=True, help="Output .xlsx path")
    args = parser.parse_args()

    with open(args.data, encoding="utf-8") as f:
        rows = json.load(f)

    create_report(args.title, args.summary, rows, args.logo, args.output)
